### PR TITLE
Expose CPU/GPU selection in Napari (default to GPU), remove use_scipy param (always True on CPU), and improve batch_size docstring (#464)

### PR DESCRIPTION
**Pull request by Fede2717**
## Description

**What is this PR**

- [ ] Bug fix
- [X] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
To expose GPU/CPU selection and batch size control in the Napari plugin, and to improve usability and performance by automatically handling the `use_scipy` setting depending on the selected device.

**What does this PR do?**
I added device selection (CPU/GPU) in the Napari plugin, defaulting to GPU if available (currently only handling "cuda").

I added some docstrings related to `batch_size`. The `batch_size` parameter was already exposed in the Napari plugin for classification.

The `use_scipy` parameter is now handled automatically: if `torch_device` is "cpu", it’s set to `True`; otherwise, `False`. It is no longer exposed in any interface, as requested.

Notes
In `cellfinder/core/main.py`, the parameter passed to the detection step (`detect`) is the one named `classification_batch_size`, while the `batch_size` parameter is used during the classification stage.

The CLI has not been modified, as it lives in the `brainglobe-workflows` repository.

Device selection currently supports `"cuda"` only. Support for other backends like `"mps"` can be added in a future update if needed.

## References

Please reference any existing issues/PRs that relate to this PR.
Issue #464

## How has this PR been tested?

All changes have been tested locally by running the test suite provided in the repository using `pytest`. All tests pass successfully. I also executed all pre-commit hooks (ruff, black, etc.) which report no issues. Additionally, SonarCloud analysis confirms that our changes meet the Quality Gate requirements. I’ll also attach a few screenshots from the tests.

## Is this a breaking change?

No, this PR does not break any existing functionality. All changes are backward-compatible.

## Does this PR require an update to the documentation?
No, only internal docstrings were updated.

## Checklist:

- [X] The code has been tested locally
- [ ] Tests have been added to cover all new functionality (unit & integration)
- [X] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by adamltyson** [comment]
Also, would you be able to take a look and see why the tests are failing?

**Comment by alessandrofelder** [comment]
I will review in-depth once you've addressed Adam's comments, Federico. Thanks for the contribution!

**Comment by Fede2717** [comment]
Thank you for your feedback, @adamltyson and @alessandrofelder!

I've updated the PR title as requested and removed the screenshots.

I noticed that some checks are failing on GitHub Actions. I'll investigate to see what's causing the failures and try to fix them as soon as possible.

Thank you again, and please let me know if there's anything else.

**Comment by alessandrofelder** [comment]
>  I'll investigate to see what's causing the failures and try to fix them as soon as possible.

Cool, I'll review when they are fixed - thanks!

**Comment by Fede2717** [comment]
@alessandrofelder, would you mind taking a look at this when you have a moment? I’m waiting on your feedback to know how best to proceed. Thanks!

File: cellfinder/core/main.py [hunk 1]
```diff
@@ -54,6 +55,11 @@ def main(
         Called with the batch number that has just finished.
     detect_finished_callback : Callable[list], optional
         Called after detection is finished with the list of detected points.
+        Number of points processed simultaneously by the neural network.
```
Comment (alessandrofelder) [inline cellfinder/core/main.py:58]: Should this be after "batch_size:"?
Or I don't understand this change?

File: cellfinder/core/main.py [hunk 2]
```diff
@@ -42,6 +42,7 @@ def main(
     detect_callback: Optional[Callable[[int], None]] = None,
     classify_callback: Optional[Callable[[int], None]] = None,
     detect_finished_callback: Optional[Callable[[list], None]] = None,
+    detection_torch_device: Optional[str] = None,
```
Comment (alessandrofelder) [inline cellfinder/core/main.py:45]: I think we just want one `torch_device` variable for both detection and classification - I don't think there is a use case where we would run detection on GPU and classification on CPU, or vice versa.

File: cellfinder/core/main.py [hunk 3]
```diff
@@ -54,6 +55,11 @@ def main(
         Called with the batch number that has just finished.
     detect_finished_callback : Callable[list], optional
         Called after detection is finished with the list of detected points.
+        Number of points processed simultaneously by the neural network.
+    batch_size: int = 64, optional
```
Comment (alessandrofelder) [inline cellfinder/core/main.py:59]:
```suggestion
    batch_size: int, optional
```
For consistency with the other parameters in the docstring, which don't list their default values.

File: cellfinder/core/main.py (hunk 1 above)
Comment (alessandrofelder) [reply cellfinder/core/main.py:58]: Also, I think it's "number of planes copied to the GPU memory at once"? There is no neural network used during detection.

File: cellfinder/napari/detect/detect.py [hunk 4]
```diff
@@ -319,6 +320,11 @@ def widget(
             Increase logging
         reset_button :
             Reset parameters to default
+        detection_torch_device : str
```
Comment (alessandrofelder) [inline cellfinder/napari/detect/detect.py:323]: I think this should be a checkbox, for a better user experience. I think it's label should be "Use GPU" and it should be ticked by default if `torch.cuda.is_available()` and greyed out if not.

**Review by alessandrofelder — CHANGES_REQUESTED** [review]
>Remove use_scipy from the failing tests and adapt them to the new logic (CPU uses SciPy, GPU uses PyTorch).

yea, this is sensible.

In general, it's about half-way there, I think. Made some suggestions.

**Comment by alessandrofelder** [comment]
Another thought: Can the CPU/GPU choice in `napari` be part of the Miscellaneous container (the `MiscInputs` class). It should appear (I think as a checkbox, see my suggestion) directly below the `analyse_local` checkbox:

Currently it shows between the reset and the run button.
image

**Comment by Fede2717** [comment]
Thanks for the feedback, @alessandrofelder. I’ll work on these changes and push an update as soon as I can.

**Comment by Fede2717** [comment]
Hi @alessandrofelder, thanks again for your detailed feedback!

Here’s what I’ve changed based on your suggestions:

Batch size docstring
In cellfinder/core/main.py, I removed the entire docstring for batch_size, because in main() only the keyword-only parameters (after *) were documented in detail.

Single torch_device
I eliminated separate detection/classification devices and unified everything into a single torch_device parameter.

Checkbox for GPU
Instead of detection_torch_device, there's now a "Use GPU" checkbox in MiscInputs (enabled by default if torch.cuda.is_available()).

Removed use_scipy in tests
As agreed, the CPU now uses SciPy filters internally and the GPU uses PyTorch. I removed references to use_scipy in the failing tests.

Changed classification_torch_device to torch_device in tests
To adapt them to the new unified logic.

I followed all of your suggestions during these updates. All tests now pass locally, and the CI checks are successful.

Please let me know if there's anything else I can improve. Otherwise, I believe this PR is ready for review/merge!
Thank you so much for your time and guidance!

**Comment by Fede2717** [comment]
Hi @alessandrofelder, just a quick friendly reminder following up on the latest updates.

I’ve implemented all the requested changes as discussed (single torch_device, GPU checkbox in the UI, docstring updates, and revised tests). The PR has been stable for about a week now, and all checks are passing.

I wanted to ask if you’ve had a chance to review the updated version once more or if there are any additional aspects you’d like me to address before merging. Please let me know if there’s anything else to adjust, or if you think it’s ready to merge.

Thanks again for your time and support!

**Review by alessandrofelder — CHANGES_REQUESTED** [review]
Thanks @Fede2717 - and sorry it took me so long to get back to you.
I am happy with what the UI looks like now, but I am worried there is a bug: I don't think there is any effect if the user ticks or unticks the `use_gpu` checkbox, because it's not connected to the underlying functionality... is that right, or am I misunderstanding?

Let me know if you're still willing to work on this, too - if not, I am happy to wrap this up for you in the next two weeks or so.

**Comment by Fede2717** [comment]
Hi @alessandrofelder, thanks for the patience and for calling this out! Sorry for the delay in getting back to you.

I’ve now updated cellfinder/napari/detect/detect.py so that in the widget() function we pass use_gpu explicitly as keyword arguments to MiscInputs.
With this change, the “Use GPU” checkbox should now correctly toggle between "cuda" and "cpu". Could you please give it another look and let me know if it’s behaving as expected or if there’s anything else I should adjust? Thanks!

**Comment by Fede2717** [comment]
Thanks so much for your review and for merging this, @alessandrofelder ! I really appreciate all your feedback and patience.
